### PR TITLE
VB-1296 Technical Debt - Improve performance and reduce complexity of…

### PR DESCRIPTION
**Pull request by AledWEvans**
… DB queries

## What does this pull request do?

Reduces complexity of DB queries and reduces unnecessary code, also performance improvements

## What is the intent behind these changes?

The intent is to make the code more maintainable and improve speed

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 1]
```diff
@@ -118,4 +116,39 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
...
+    @Param("prisonId") prisonId: String,
+    @Param("startDateTime") startDateTime: LocalDateTime? = null,
+    @Param("endDateTime") endDateTime: LocalDateTime? = null,
+    @Param("visitStatus") visitStatus: VisitStatus? = null,
+    @Param("nomisPersonId") nomisPersonId: Long? = null,
+    page: Pageable
+  ): Page<Visit>
+
+  @Query(
+    "SELECT CASE WHEN (COUNT(v) > 0) THEN TRUE ELSE FALSE END FROM Visit v LEFT JOIN v.visitors as vis " +
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:140]: need a integration test to cover hasVisits , as the unit tests just mock it

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:140]: added new test, to improve coverage

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt [hunk 2]
```diff
@@ -402,25 +399,24 @@ class VisitController(
     @Parameter(
       description = "Pagination page number, starting at zero",
       example = "0"
-    ) page: Int?,
+    ) page: Int,
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt:402]: I will have a look into this

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt [hunk 3]
```diff
@@ -328,14 +327,12 @@ class VisitController(
...
     visitService.findVisitsByFilter(
-      VisitFilter(
-        prisonerId = prisonerId?.trim(),
-        prisonId = prisonId?.trim(),
-        startDateTime = startTimestamp,
-        endDateTime = endTimestamp,
-        nomisPersonId = nomisPersonId,
-        visitStatus = visitStatus
-      )
+      prisonerId = prisonerId,
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt:330]: trim in service 
    return findVisitsByFilterPageableDescending(
```
      prisonerId = prisonerId?.trim(),
      prisonId = prisonId.trim(),
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 4]
```diff
@@ -118,4 +116,39 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
     nativeQuery = true
   )
   fun updateCreateTimestamp(createTimestamp: LocalDateTime, id: Long): Int
+
+  @Query(
+    "SELECT v FROM Visit v LEFT JOIN v.visitors as vis " +
+      "WHERE (:visitStatus is null OR v.visitStatus = :visitStatus)  AND " +
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:122]: This is the only bit VisitSpecification was better at handling,  But the result should be the same..

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt]: Have moved this out of service and back in controller to fix data at earliest opportunity,.. this is best practice

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt:402]: Agreed after the call we had yesterday to discuss.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:122]: Agreed after Aled demoed how these queries work. I was of the opinion that we are checking for visitStatus being NULL in DB but that is not the case and this query works well.

**Pull request merged**